libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/AbbreviationCoordinator.java

```diff
@@ -0,0 +1,145 @@
+package com.mapbox.services.android.navigation.ui.v5.instruction;
+
+import android.widget.TextView;
+
+import com.mapbox.api.directions.v5.models.BannerComponents;
+import com.mapbox.services.android.navigation.ui.v5.instruction.InstructionLoader.BannerComponentNode;
+import com.mapbox.services.android.navigation.ui.v5.utils.TextViewUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class allows text to be constructed to fit a given TextView, given specified
+ * BannerComponents containing abbreviation information and given a list of BannerComponentNodes,
+ * constructed by InstructionLoader.
+ */
+public class AbbreviationCoordinator {
+  private static final String SINGLE_SPACE = " ";
+  private Map<Integer, List<Integer>> abbreviations;
+  private TextViewUtils textViewUtils;
+
+  public AbbreviationCoordinator(TextViewUtils textViewUtils) {
+    this.abbreviations = new HashMap<>();
+    this.textViewUtils = textViewUtils;
+  }
+
+  public AbbreviationCoordinator() {
+    this(new TextViewUtils());
+  }
+
+  /**
+   * Adds the given BannerComponents object to the list of abbreviations so that when the list of
+   * BannerComponentNodes is completed, text can be abbreviated properly to fit the specified
+   * TextView.
+   *
+   * @param bannerComponents object holding the abbreviation information
+   * @param index in the list of BannerComponentNodes
+   */
+  public void addPriorityInfo(BannerComponents bannerComponents, int index) {
+    int abbreviationPriority = bannerComponents.abbreviationPriority();
+    if (abbreviations.get(Integer.valueOf(abbreviationPriority)) == null) {
+      abbreviations.put(abbreviationPriority, new ArrayList<Integer>());
+    }
+    abbreviations.get(abbreviationPriority).add(index);
+  }
+
+  /**
+   * Using the abbreviations HashMap which should already be populated, abbreviates the text in the
+   * bannerConmponentNodes until the text fits the given TextView.
+   *
+   * @param bannerComponentNodes containing the text to construct
+   * @param textView to check the text fits
+   * @return the properly abbreviated string that will fit in the TextView
+   */
+  public String abbreviateBannerText(List<BannerComponentNode> bannerComponentNodes, TextView textView) {
+    String bannerText = join(bannerComponentNodes);
+
+    if (abbreviations.isEmpty()) {
+      return bannerText;
+    }
+
+    int currAbbreviationPriority = 0;
+    int maxAbbreviationPriority = Collections.max(abbreviations.keySet());
+
+    while (shouldKeepAbbreviating(textView, bannerText, currAbbreviationPriority, maxAbbreviationPriority)) {
```

Comment:
@Guardiola31337 I extracted it into a method because otherwise there would be duplicate code (before the while loop and at the end of it). And you're saying extract the whole while loop?

Reply:
Correct. I guess that block of code could be extracted into a `private` method. This way `addPriorityInfo` will be easier to read and understand and comment will become superfluous so it won't be necessary. The idea is to keep `public` methods as clean as possible so they read like pseudocode (i.e. no conditionals, `for`s, etc.). Does that make sense?